README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->.

# 时间轴的widget


## Features
![img_1.png](img_1.png)
用于预约会议室时间段显示

## Getting started

```dart
dependencies:
  flutter:
    sdk: flutter
  time_slot_widget: 0.1.0
```

## Usage

```dart
TimeSlotWidget(
  timeItemBuilder: (int index) {
    int start = 10 + index * max(2, indexFirst + 2);
    int end = start + 1;
    return TimeItemEntity(
      start: buildDateTime(start),
       end: buildDateTime(end),
    );
  },
  count: 3,
  timeStyle: TimeStyle(
    height: (50 + indexFirst).toDouble(),
    timeHeight: (24 + indexFirst*2).toDouble(),
  ),
);

```

## Additional information




CHANGELOG.md
## 0.1.1

* 更新配置

## 0.1.0

* init


test/time_slot_widget_test.dart
// import 'package:flutter_test/flutter_test.dart';
//
// import 'package:time_slot_widget/time_slot_widget.dart';
//
// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//   });
// }


flutter_service_worker.js
'use strict';
const MANIFEST = 'flutter-app-manifest';
const TEMP = 'flutter-temp-cache';
const CACHE_NAME = 'flutter-app-cache';
const RESOURCES = {
  "version.json": "ff966ab969ba381b900e61629bfb9789",
"index.html": "24a4f01846e22017a8b69bb7d1ad5452",
"/": "24a4f01846e22017a8b69bb7d1ad5452",
"main.dart.js": "c4a11af7ce38404e2ad78356cdaaa6db",
"flutter.js": "f85e6fb278b0fd20c349186fb46ae36d",
"favicon.png": "5dcef449791fa27946b3d35ad8803796",
"icons/Icon-192.png": "ac9a721a12bbc803b44f645561ecb1e1",
"icons/Icon-maskable-192.png": "c457ef57daa1d16f64b27b786ec2ea3c",
"icons/Icon-maskable-512.png": "301a7604d45b3e739efc881eb04896ea",
"icons/Icon-512.png": "96e752610906ba2a93c65f8abe1645f1",
"manifest.json": "0867c3e13649ac4d06fe34b7b3ddce08",
"assets/AssetManifest.json": "2efbb41d7877d10aac9d091f58ccd7b9",
"assets/NOTICES": "0da80773cc06b0fac966d250e214dca4",
"assets/FontManifest.json": "dc3d03800ccca4601324923c0b1d6d57",
"assets/packages/cupertino_icons/assets/CupertinoIcons.ttf": "6d342eb68f170c97609e9da345464e5e",
"assets/shaders/ink_sparkle.frag": "fe5ae851a8451bcb83bb274b269caa4d",
"assets/fonts/MaterialIcons-Regular.otf": "95db9098c58fd6db106f1116bae85a0b",
"canvaskit/canvaskit.js": "2bc454a691c631b07a9307ac4ca47797",
"canvaskit/profiling/canvaskit.js": "38164e5a72bdad0faa4ce740c9b8e564",
"canvaskit/profiling/canvaskit.wasm": "95a45378b69e77af5ed2bc72b2209b94",
"canvaskit/canvaskit.wasm": "bf50631470eb967688cca13ee181af62"
};

// The application shell files that are downloaded before a service worker can
// start.
const CORE = [
  "main.dart.js",
"index.html",
"assets/AssetManifest.json",
"assets/FontManifest.json"];
// During install, the TEMP cache is populated with the application shell files.
self.addEventListener("install", (event) => {
  self.skipWaiting();
  return event.waitUntil(
    caches.open(TEMP).then((cache) => {
      return cache.addAll(
        CORE.map((value) => new Request(value, {'cache': 'reload'})));
    })
  );
});

// During activate, the cache is populated with the temp files downloaded in
// install. If this service worker is upgrading from one with a saved
// MANIFEST, then use this to retain unchanged resource files.
self.addEventListener("activate", function(event) {
  return event.waitUntil(async function() {
    try {
      var contentCache = await caches.open(CACHE_NAME);
      var tempCache = await caches.open(TEMP);
      var manifestCache = await caches.open(MANIFEST);
      var manifest = await manifestCache.match('manifest');
      // When there is no prior manifest, clear the entire cache.
      if (!manifest) {
        await caches.delete(CACHE_NAME);
        contentCache = await caches.open(CACHE_NAME);
        for (var request of await tempCache.keys()) {
          var response = await tempCache.match(request);
          await contentCache.put(request, response);
        }
        await caches.delete(TEMP);
        // Save the manifest to make future upgrades efficient.
        await manifestCache.put('manifest', new Response(JSON.stringify(RESOURCES)));
        return;
      }
      var oldManifest = await manifest.json();
      var origin = self.location.origin;
      for (var request of await contentCache.keys()) {
        var key = request.url.substring(origin.length + 1);
        if (key == "") {
          key = "/";
        }
        // If a resource from the old manifest is not in the new cache, or if
        // the MD5 sum has changed, delete it. Otherwise the resource is left
        // in the cache and can be reused by the new service worker.
        if (!RESOURCES[key] || RESOURCES[key] != oldManifest[key]) {
          await contentCache.delete(request);
        }
      }
      // Populate the cache with the app shell TEMP files, potentially overwriting
      // cache files preserved above.
      for (var request of await tempCache.keys()) {
        var response = await tempCache.match(request);
        await contentCache.put(request, response);
      }
      await caches.delete(TEMP);
      // Save the manifest to make future upgrades efficient.
      await manifestCache.put('manifest', new Response(JSON.stringify(RESOURCES)));
      return;
    } catch (err) {
      // On an unhandled exception the state of the cache cannot be guaranteed.
      console.error('Failed to upgrade service worker: ' + err);
      await caches.delete(CACHE_NAME);
      await caches.delete(TEMP);
      await caches.delete(MANIFEST);
    }
  }());
});

// The fetch handler redirects requests for RESOURCE files to the service
// worker cache.
self.addEventListener("fetch", (event) => {
  if (event.request.method !== 'GET') {
    return;
  }
  var origin = self.location.origin;
  var key = event.request.url.substring(origin.length + 1);
  // Redirect URLs to the index.html
  if (key.indexOf('?v=') != -1) {
    key = key.split('?v=')[0];
  }
  if (event.request.url == origin || event.request.url.startsWith(origin + '/#') || key == '') {
    key = '/';
  }
  // If the URL is not the RESOURCE list then return to signal that the
  // browser should take over.
  if (!RESOURCES[key]) {
    return;
  }
  // If the URL is the index.html, perform an online-first request.
  if (key == '/') {
    return onlineFirst(event);
  }
  event.respondWith(caches.open(CACHE_NAME)
    .then((cache) =>  {
      return cache.match(event.request).then((response) => {
        // Either respond with the cached resource, or perform a fetch and
        // lazily populate the cache only if the resource was successfully fetched.
        return response || fetch(event.request).then((response) => {
          if (response && Boolean(response.ok)) {
            cache.put(event.request, response.clone());
          }
          return response;
        });
      })
    })
  );
});

self.addEventListener('message', (event) => {
  // SkipWaiting can be used to immediately activate a waiting service worker.
  // This will also require a page refresh triggered by the main worker.
  if (event.data === 'skipWaiting') {
    self.skipWaiting();
    return;
  }
  if (event.data === 'downloadOffline') {
    downloadOffline();
    return;
  }
});

// Download offline will check the RESOURCES for all files not in the cache
// and populate them.
async function downloadOffline() {
  var resources = [];
  var contentCache = await caches.open(CACHE_NAME);
  var currentContent = {};
  for (var request of await contentCache.keys()) {
    var key = request.url.substring(origin.length + 1);
    if (key == "") {
      key = "/";
    }
    currentContent[key] = true;
  }
  for (var resourceKey of Object.keys(RESOURCES)) {
    if (!currentContent[resourceKey]) {
      resources.push(resourceKey);
    }
  }
  return contentCache.addAll(resources);
}

// Attempt to download the resource online before falling back to
// the offline cache.
function onlineFirst(event) {
  return event.respondWith(
    fetch(event.request).then((response) => {
      return caches.open(CACHE_NAME).then((cache) => {
        cache.put(event.request, response.clone());
        return response;
      });
    }).catch((error) => {
      return caches.open(CACHE_NAME).then((cache) => {
        return cache.match(event.request).then((response) => {
          if (response != null) {
            return response;
          }
          throw error;
        });
      });
    })
  );
}


lib/time_slot_widget.dart
library time_slot_widget;

import 'dart:math';
import 'dart:ui' as ui;

import 'package:flutter/material.dart';

/// 自定义时间区块，构建指定格式的实体
typedef TimeItemBuilder = TimeItemEntity Function(int index);

/// 时间轴
class TimeSlotWidget extends StatefulWidget {
  /// 控制当前实际时间
  final DateTime? currentDateTime;

  ///  时间段的总数
  final int? count;

  /// 时间段自定义
  final TimeItemBuilder timeItemBuilder;

  /// 各项参数配置
  final TimeStyle? timeStyle;

  const TimeSlotWidget({
    Key? key,
    required this.timeItemBuilder,
    this.currentDateTime,
    this.count,
    this.timeStyle,
  }) : super(key: key);

  @override
  TimeSlotWidgetState createState() => TimeSlotWidgetState();
}

class TimeSlotWidgetState extends State<TimeSlotWidget> {
  TimeStyle get _timeStyle {
    return widget.timeStyle ?? TimeStyle();
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(builder: (context, BoxConstraints boxConstraints) {
      return SizedBox(
        height: _timeStyle.height,
        width: double.infinity,
        child: CustomPaint(
          painter: _TimeZonePainter(
            boxConstraints.maxWidth,
            widget.currentDateTime ?? DateTime.now(),
            timeStyle: _timeStyle,
            itemCount: widget.count,
            timeItemBuilder: widget.timeItemBuilder,
          ),
        ),
      );
    });
  }
}

class _TimeZonePainter extends CustomPainter with RatioFromTimeMixin {
  final TimeItemBuilder timeItemBuilder;
  final double width;
  final int? itemCount;
  final TimeStyle timeStyle;

  _TimeZonePainter(
      this.width,
      DateTime currentDateTime, {
        required this.timeItemBuilder,
        required this.timeStyle,
        this.itemCount,
      }) : super() {
    bgPainter = Paint()..strokeWidth = 1;
    this.currentDateTime = currentDateTime;
    start = timeStyle.start;
    end = timeStyle.end;
  }

  late Paint bgPainter;

  /// 计算背景显示区块数
  int get allCount {
    return ((timeStyle.end - timeStyle.start).abs() / timeStyle.space).floor();
  }

  @override
  void paint(Canvas canvas, Size size) {
    /// 绘制背景
    drawBg(canvas, size, timeStyle.bgFillColor);

    /// 绘制时间线刻度
    drawBgZone(canvas, size, timeStyle.timeLineFillColor);

    /// 绘制刻度对应的数字
    drawNumber(canvas, size, timeStyle.numberTextColor);

    /// 绘制需要显示的内容
    drawTime(canvas, size);

    /// 绘制当前时间
    drawCurrentTime(canvas, size);
  }

  /// 指定颜色的背景
  void drawBg(Canvas canvas, Size size, Color bgColor) {
    Rect zone1 = Rect.fromLTWH(0, 0, size.width, size.height);
    bgPainter
      ..style = PaintingStyle.fill
      ..color = bgColor;
    canvas.drawRect(zone1, bgPainter);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }

  void drawBgZone(Canvas canvas, Size size, Color color) {
    int i = 0;
    Paint painter = Paint()
      ..strokeWidth = 1
      ..color = color
      ..style = PaintingStyle.fill;
    int num = allCount;
    double itemWidth = width / allCount;
    while (i < num) {
      Rect zone1 = Rect.fromLTWH(itemWidth * i, 0,
          itemWidth - timeStyle.timeSpace, timeStyle.timeHeight);
      canvas.drawRect(zone1, painter);
      i++;
    }
  }

  void drawNumber(Canvas canvas, Size size, Color textColor) {
    int i = 1;
    int num = allCount;
    double itemWidth = width / num;
    while (i < num) {
      _paintText(
        canvas,
        size,
        Offset(itemWidth * i, timeStyle.timeHeight + timeStyle.numberTop),
        info: (timeStyle.start + timeStyle.space * i).toString(),
        textColor: textColor,
        fontSize: timeStyle.numberFontSize,
        textMaxWidth: timeStyle.textMaxWidth,
      );
      i++;
    }
  }

  // 绘制文字
  void _paintText(
      Canvas canvas,
      Size size,
      Offset offset, {
        required String info,
        double textMaxWidth = 20,
        TextAlign? textAlign,
        double? fontSize,
        Color? textColor,
      }) {
    ui.ParagraphBuilder pb = ui.ParagraphBuilder(
      ui.ParagraphStyle(
        textAlign: textAlign ?? TextAlign.center,
        fontWeight: FontWeight.normal,
        fontStyle: FontStyle.normal,
        fontSize: fontSize ?? 12,
        height: 1,
        maxLines: 1,
      ),
    )
      ..pushStyle(ui.TextStyle(color: textColor ?? const Color(0xFF999999)))
      ..addText(info);
    ui.ParagraphConstraints pc = ui.ParagraphConstraints(width: textMaxWidth);
    ui.Paragraph paragraph = pb.build()..layout(pc);
    canvas.drawParagraph(
        paragraph, Offset(offset.dx - textMaxWidth / 2, offset.dy));
  }

  void drawTime(ui.Canvas canvas, ui.Size size) {
    if (itemCount == null || itemCount == 0) {
      return;
    }
    int count = itemCount!;
    for (int i = 0; i < count; i++) {
      TimeItemEntity timeItemEntity = timeItemBuilder.call(i);
      double start = offsetFromDateTime(timeItemEntity.start);
      double end = offsetFromDateTime(timeItemEntity.end);
      Paint painter = Paint()
        ..strokeWidth = 1
        ..color = timeItemEntity.fillColor
        ..style = PaintingStyle.fill;
      Rect zone = Rect.fromLTWH(
        width * start,
        0,
        width * (end - start),
        timeStyle.timeHeight,
      );
      canvas.drawRect(zone, painter);
    }
  }

  void drawCurrentTime(
      Canvas canvas,
      Size size, {
        double dashHeight = 4,
        double dashSpace = 2,
        double startY = 0.5,
      }) {
    if (!DateUtils.isSameDay(DateTime.now(), currentDateTime)) {
      return;
    }
    double begin = offsetFromDateTime(DateTime.now());
    var paint = Paint() // 创建一个画笔并配置其属性
      ..strokeWidth = 1 // 画笔的宽度
      ..isAntiAlias = true // 是否抗锯齿
      ..color = const Color(0xFFFF4A4A); // 画笔颜色
    var max = timeStyle.timeHeight;
    double startX = width * begin;
    int count = 0;
    while (startY < max) {
      canvas.drawLine(
          Offset(startX, startY),
          Offset(
              startX,
              min(
                  count == 0
                      ? startY + timeStyle.dashLineSeparate
                      : startY + dashHeight,
                  max)),
          paint);
      startY +=
          (count == 0 ? timeStyle.dashLineSeparate : dashHeight) + dashSpace;
      count++;
    }
  }
}

mixin RatioFromTimeMixin {
  late DateTime currentDateTime;
  late int start;
  late int end;

  /// 计算时间所占UI的距离左侧的偏移量，最大为1，0-1范围内
  double offsetFromDateTime(DateTime fromTime) {
    int all = maxTime.difference(minTime).inMinutes;
    return correctDuration(fromTime, minTime).inMinutes / all;
  }

  DateTime get minTime {
    return DateTime(currentDateTime.year, currentDateTime.month,
        currentDateTime.day, start, 00, 00);
  }

  DateTime get maxTime {
    return DateTime(currentDateTime.year, currentDateTime.month,
        currentDateTime.day, end, 00, 00);
  }

  Duration correctDuration(DateTime dateTime, DateTime minDateTime) {
    Duration endDuration = dateTime.difference(minDateTime);

    if (endDuration.inMinutes < 0) {
      endDuration = const Duration(minutes: 0);
    }
    return endDuration;
  }
}

class TimeItemEntity {
  /// 填充颜色
  final Color fillColor;

  /// 开始时间
  final DateTime start;

  /// 结束时间
  final DateTime end;

  TimeItemEntity({
    this.fillColor = const Color(0xffDDDDDD),
    required this.start,
    required this.end,
  });
}

class TimeStyle {
  /// 背景色，默认透明色
  final Color bgFillColor;

  /// 刻度色块
  final Color timeLineFillColor;

  /// 记录的时间区块的颜色
  final Color showTimeLineFillColor;

  /// 刻度对应的文字的颜色
  final Color numberTextColor;

  /// 限制时间开始，默认8点
  final int start;

  /// 限制时间结束，默认24点
  final int end;

  /// 限制时间 时间间隔，默认2小时
  final int space;

  /// 整体高度 默认50
  final double height;

  /// 时间区块的高度，默认24
  final double timeHeight;

  /// 时间区块的间隔 默认1.62
  final double timeSpace;

  /// 刻度文字距离上间距，默认8
  final double numberTop;

  /// 刻度文字的字体大小 默认12
  final double numberFontSize;

  /// 刻度文字的最大宽度 默认20；可以根据世界内容调整
  final double textMaxWidth;

  /// 虚线间隔间距
  final double dashLineSeparate;

  TimeStyle({
    this.bgFillColor = Colors.transparent,
    this.timeLineFillColor = const Color(0xfff2f2f2),
    this.showTimeLineFillColor = const Color(0xFFDDDDDD),
    this.numberTextColor = const Color(0xFF999999),
    this.height = 50,
    this.timeHeight = 24,
    this.timeSpace = 1.62,
    this.start = 8,
    this.end = 24,
    this.space = 2,
    this.numberTop = 8,
    this.numberFontSize = 12,
    this.textMaxWidth = 20,
    this.dashLineSeparate = 2,
  });
}


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:time_slot_widget/time_slot_widget.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'time slot widget',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'time slot widget'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        padding: const EdgeInsets.all(16),
        child: ListView.separated(
          itemBuilder: (BuildContext context, int indexFirst) {
            return TimeSlotWidget(
              timeItemBuilder: (int index) {
                int start = 10 + index * max(2, indexFirst + 2);
                int end = start + 1;
                return TimeItemEntity(
                  start: buildDateTime(start),
                  end: buildDateTime(end),
                  fillColor: [Colors.red,Colors.amber,Colors.black][Random().nextInt(3)],
                );
              },
              count: 3,
              timeStyle: TimeStyle(
                height: (50 + indexFirst).toDouble(),
                timeHeight: (24 + indexFirst*2).toDouble(),
                numberTextColor: [Colors.red,Colors.amber,Colors.black][Random().nextInt(3)],
              ),
            );
          },
          separatorBuilder: (BuildContext context, int index) {
            return const SizedBox(height: 10);
          },
          itemCount: 5,
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }

  DateTime buildDateTime(int hour, {int minutes = 0}) {
    DateTime now = DateTime.now();
    return DateTime(now.year, now.month, now.day, hour, minutes, 0);
  }
}
